packages for managing clinical trial data (#273)
pharmaverse/blog: posts/2025-02-17_theres_a_pharmaverse_package_for_that/managing-clinical-trial-data.qmd

This post covers:

-   Key stages of clinical trials and the R packages that support them

-   Creating ADSL datasets and essential programming steps

-   Key players in pharmaverse and whether you need all packages

-   How pharmaverse compares to Tidyverse and how to learn it

By the end, you'll have a clear understanding of how pharmaverse supports clinical trial operations and how to apply these tools in your work.

## **Key Stages of Clinical Reporting**

Managing clinical trial data involves multiple stages, each with its own challenges. **Pharmaverse** provides a range of R packages that support different parts of the process, sometimes even offering multiple options for the same task. This flexibility allows organizations to choose the best tools for their specific needs rather than sticking to a one-size-fits-all approach.

A **metadata-driven approach** helps ensure that clinical trial data is consistently structured and aligned with regulatory standards. The typical process follows this sequence:

**Metadata** ➝ **OAK** ➝ **Admiral** ➝ **Define.xml** ➝ **TLGs** ➝ **Submissions**

Some examples of **pharmaverse** packages that support clinical reporting include:

> -   [**diffdf**](https://cran.r-project.org/web/packages/diffdf/vignettes/diffdf-basic.html) – Tracking differences in datasets.

Author: I don't understand this either please. 😅

> -   [**diffdf**](https://cran.r-project.org/web/packages/diffdf/vignettes/diffdf-basic.html) – Tracking differences in datasets.

Reviewer: the links package automatically links to the pkgdown site for R packages that use this `{admiral}` in the post.  You can see previous couple of posts on how we do this.

-   [**metatools**](https://pharmaverse.github.io/metatools/) – Metadata management and transformation.
-   [**sdtm.oak**](https://pharmaverse.github.io/sdtm.oak/) – The primary **pharmaverse** package for SDTM dataset creation.
-   [**sdtmchecks**](https://pharmaverse.github.io/sdtmchecks/) – Validates SDTM datasets for compliance with CDISC standards.
-   [**admiral**](https://github.com/pharmaverse/admiral) – Standardized data derivations.
-   [**metacore**](https://cran.r-project.org/web/packages/metacore/index.html) – Metadata-driven structures.
-   [**rtables**](https://github.com/insightsengineering/rtables) – Structured table outputs.

> -   The **pharmaverse** provides multiple table-making packages, such as [chevron](https://insightsengineering.github.io/chevron/latest-tag/), [Tplyr](https://github.com/atorus-research/Tplyr), [pharmaRTF](https://atorus-research.github.io/pharmaRTF/), [gtsummary](https://www.danieldsjoberg.com/gtsummary/), [cards](https://insightsengineering.github.io/cards/latest-tag/), [tfrmt](https://gsk-biostatistics.github.io/tfrmt/), and [tidytlg](https://pharmaverse.github.io/tidytlg/main/). More tools are listed on the [TLGs page](https://pharmaverse.org/e2eclinical/tlg/).

Author: Is it worth mentioning rtables here again as I had included it before but asked to take it out?  Maybe I don't understand your suggestion.

> -   The **pharmaverse** provides multiple table-making packages, such as [chevron](https://insightsengineering.github.io/chevron/latest-tag/), [Tplyr](https://github.com/atorus-research/Tplyr), [pharmaRTF](https://atorus-research.github.io/pharmaRTF/), [gtsummary](https://www.danieldsjoberg.com/gtsummary/), [cards](https://insightsengineering.github.io/cards/latest-tag/), [tfrmt](https://gsk-biostatistics.github.io/tfrmt/), and [tidytlg](https://pharmaverse.github.io/tidytlg/main/). More tools are listed on the [TLGs page](https://pharmaverse.org/e2eclinical/tlg/).

Reviewer 2: Ah alright, so: 
chevron is designed to create standard outputs. As an engine for it's tables it uses the rtables package - but it can create also graphs (using tern, iirc) and listings (using rlistings). Essentially, chevron is a wrapper package.
I think it would be useful to group these together, as they are jointly developed. 
Whereas Tplyr is a different package with a different approach. There may be a wrapper package for Tplyr to create standard outputs as well, but I'm not aware of this.

---
title: "Managing Clinical Trial Data? There’s a Pharmaverse Package for That"
author:
  - name: Gift Kenneth
  - name: Sunil Gupta
  - name: APPSILON

> description: "Looking for R packages to manage clinical trial data? pharmaverse has tools for every stage from data collection to submission."

Reviewer 3: ```suggestion
description: "Looking for R packages to manage clinical trial data? Pharmaverse has tools for every stage from data collection to submission."
```

we generally capitalise pharmaverse at the start of sentences

long_slug <- "2025-02-17_managing_clinical_trial_data..."
# renv::use(lockfile = "renv.lock")
```

<!--------------- post begins here ----------------->

Managing clinical trial data is no small task. It needs to be precise, compliant, and efficient. Traditionally, this meant using proprietary tools and working within siloed systems, which often made the process more complicated and expensive than necessary. But we think there’s a better way.

The [**pharmaverse**](https://pharmaverse.org/) is an open-source ecosystem of R packages built specifically for clinical trials. These tools integrate seamlessly with the [Tidyverse](https://www.tidyverse.org/), making data management more flexible, efficient, and transparent.

Whether you’re collecting, validating, analyzing, or preparing data for regulatory submission, there’s a pharmaverse package designed to support your workflow and help you work smarter.

This post covers:

-   Key stages of clinical trials and the R packages that support them

-   Getting started with metadata-driven clinical reporting

-   The process from raw data to ADaMs, including validation

-   Creating ADSL datasets and essential programming steps

> -   Key players in Pharmaverse and whether you need all packages

Reviewer 3: ```suggestion
-   Key players in pharmaverse and whether you need all packages
```

-   How pharmaverse compares to Tidyverse and how to learn it

> By the end, you'll have a clear understanding of how Pharmaverse supports clinical trial operations and how to apply these tools in your work.

Reviewer 3: ```suggestion
By the end, you'll have a clear understanding of how pharmaverse supports clinical trial operations and how to apply these tools in your work.
```

## **Key Stages of the Clinical Trial Process**

> Managing clinical trial data involves multiple stages, each with its own challenges. pharmaverse provides a range of packages for different parts of the process, sometimes even more than one option for the same task, so you can choose what works best for your needs rather than sticking to a single approach

Reviewer 3: ```suggestion
Managing clinical trial data involves multiple stages, each with its own challenges. Pharmaverse provides a range of packages for different parts of the process, sometimes even more than one option for the same task, so you can choose what works best for your needs rather than sticking to a single approach.
```

![](images/clinical-trial-process.png)

> 

Reviewer 3: Add a sentence (preferably in bold or in a highlighted box) clarifying here that you put together this list for the blog, and if any package was missed, we are happy to update the post to include it.  I wouldn't want anyone's package to feel left out.

Reviewer 4: Or just say that these are some examples of pharmaverse packages and consult the site for the full list, as otherwise you're going to be maintaining this blog every time new packages get added

1.  **Collection** – Initial data handling using:

    -   [**tidyverse**](https://www.tidyverse.org/) for general data manipulation.

    -   [**diffdf**](https://cran.r-project.org/web/packages/diffdf/vignettes/diffdf-basic.html) for tracking differences in datasets.

>     -   [**metatools**](https://pharmaverse.github.io/metatools/) for metadata management and transformation.

Reviewer 4: I don't think these 3 are related to collection, they're more like utilities that could be used across many areas of clinical reporting. I think that's why they were added to that visual when we first created it - it wasn't saying that they're packages to help data collection

2.  **Tabulation** – Structuring and validating data with:

    -   [**thevalidatoR**](https://github.com/marketplace/actions/r-package-validation-report) for data validation checks.

>     -   [**covtracer**](https://github.com/Genentech/covtracer) for ensuring code traceability and reproducibility.

Reviewer 4: Same comment as above - these are general validation tools - not related to tabulation. I wouldn't even call these out here as the focus of your blog is clinical trial analyses whereas these are more like developer tools

Reviewer 4: The main sdtm package of pharmaverse is sdtm.oak - the visual you have was created before this one got open sourced

Reviewer 4: Also we have sdtmchecks here in this space, but as per my comment above i don't think we need to call out all - i would streamline this list and just say they're some examples of the packages offered

3.  **Analysis Ready** – Preparing data for regulatory-grade analytics:

    -   [**admiral**](https://github.com/pharmaverse/admiral) for standardized data derivations.

    -   [**metacore**](https://cran.r-project.org/web/packages/metacore/index.html) for metadata-driven structures.

    -   [**dplyr**](https://dplyr.tidyverse.org/) (part of the tidyverse) for data manipulation

4.  **Analysis Results** – Incorporating visualization and modeling tools:

    -   [**ggplot2**](https://ggplot2.tidyverse.org/) (part of the tidyverse) for advanced data visualization.

>     -   [**visR**](https://github.com/openpharma/visR) for survival analysis and Kaplan-Meier plots.

Reviewer 4: visR got removed from pharmaverse: https://github.com/pharmaverse/pharmaverse/issues/273

    -   [**pharmaRTF**](https://cran.r-project.org/web/packages/pharmaRTF/index.html) for creating submission-ready reports.

>     -   [**rtables**](https://github.com/insightsengineering/rtables) for structured table outputs

Reviewer 3: I would remove rtables here if mentioned below

Reviewer 4: also could move pharmaRTF and Tplyr down to that list of table packages rather than having their own bullets

    -   [**Tplyr**](https://github.com/atorus-research/Tplyr) for clinical summary tables

>     -   The pharmaverse offers multiple table-making packages, including [rtables](https://insightsengineering.github.io/rtables/v0.6.10/articles/rtables.html), [chevron](https://insightsengineering.github.io/chevron/latest-tag/), [pharmaRTF](https://atorus-research.github.io/pharmaRTF/), [gtsummary](https://www.danieldsjoberg.com/gtsummary/), [cards](https://insightsengineering.github.io/cards/latest-tag/), [tfrmt](https://gsk-biostatistics.github.io/tfrmt/), [tidytlg](https://pharmaverse.github.io/tidytlg/main/), and [cardinal](https://pharmaverse.github.io/cardinal/). You can find more tools on the [TLGs page](https://pharmaverse.org/e2eclinical/tlg/). Each serves different needs, and it is up to the programmer to choose the best fit.

Reviewer 4: i wouldn't mention cardinal here as its an examples book to show how to use different table packages for the FDA templates - its not a table-making package itself

5.  **Submission** – Final reporting and submission:

    -   [**xportr**](https://github.com/atorus-research/xportr) for CDISC-compliant dataset export.

    -   [**pkglite**](https://merck.github.io/pkglite/) for package management and tracking.

By using these tools, organizations can optimize their data pipeline, ensuring clinical data is well-structured, validated, and ready for regulatory submission with ease.

> ## **Where Do I Start? Understanding Tasks & Objectives**

Reviewer 3: Feels like this section is superfluous because more packages are being listed here when in the section previous to this one, a list was provided for all stages of clinical trial development. i'd remove this section and absorb key content into the previous one.

For those new to clinical reporting in R, understanding the foundational tasks and objectives is critical.

#### **Metadata-Driven Clinical Reporting Approach**

A metadata-driven method ensures clinical trial data is consistently structured and aligned with regulatory standards. The process typically follows this sequence:

> [**Metadata**](https://pharmaverse.org/e2eclinical/metadata/) **\> [OAK](https://github.com/pharmaverse/sdtm.oak) \> [Admiral](https://github.com/pharmaverse/admiral) \> Define.xml \> [TLGs](https://pharmaverse.org/e2eclinical/tlg/) (rtf/pdf) \> Submissions \> Shiny**

Reviewer 3: Question: does shiny only fit in at the end? In my experience shiny is used as a tool throughout

Reviewer 4: agree

> [**Metadata**](https://pharmaverse.org/e2eclinical/metadata/) **\> [OAK](https://github.com/pharmaverse/sdtm.oak) \> [Admiral](https://github.com/pharmaverse/admiral) \> Define.xml \> [TLGs](https://pharmaverse.org/e2eclinical/tlg/) (rtf/pdf) \> Submissions \> Shiny**

Reviewer 4: OAK -> sdtm.oak and I would remove mention of rtf/pdf as actually the TLGs can be produced in many different formats dependent on the package used

This structured approach makes data preparation for regulatory submission more transparent and traceable.

> #### **Essential Resources & Packages**

Reviewer 3: as mentioned above: i'd absorb key elements of this list into the above list.

Reviewer 4: yes agree - i find it confusing having 2 separate lists of select pharmaverse packages here. i would have one streamlined list with just a few examples, and encouraging people to go to our site to see the full offerings

To effectively implement this process, organizations should become familiar with essential packages:

-   **Metadata Management:** [metacore](https://github.com/atorus-research/metacore), [metatools](https://pharmaverse.github.io/metatools/) – for standardized metadata structures and validation.

> -   [**SDTM (Standard Data Tabulation Model)**](https://pharmaverse.org/e2eclinical/sdtm/)**:** Ensures clinical data follows CDISC standards.

Reviewer 3: this is not a package so I don't think it would fit here.

> -   **Validation Tools:** [sdtmchecks](https://github.com/pharmaverse/sdtmchecks), [datacutr](https://cran.r-project.org/web/packages/datacutr/index.html) (ensuring data integrity and regulatory compliance).

Reviewer 4: I wouldn't call these validation tools - sdtmchecks is a data checker that informs the stat programmer of any watchouts in the data that might impact analyses and datacutr is to apply a datacut to a data snapshot. I wouldn't consider either as "essential tools" as not all companies would use these, so i would remove this

For more details, explore:

-   **Clinical Metadata Resources:** <https://pharmaverse.org/e2eclinical/metadata>

-   **SDTM Standards & Tools:** <https://pharmaverse.org/e2eclinical/sdtm> 

> ## **Example: Creating ADSL**

Reviewer 4: Should we just advertise the examples site here and https://pharmaverse.github.io/examples/adam/adsl.html rather than repeating this within the blog?

Building an ADSL dataset involves several key steps, from reading in data to deriving treatment variables and population flags. While these steps apply regardless of the tools used, **pharmaverse packages like {admiral} simplify the process** with functions designed for CDISC-compliant datasets.

#### **Step 1: Read in Data**

> To begin, clinical trial datasets such as **DM, EX, DS, AE, and LB** are loaded. The {pharmaversesdtm} package provides sample CDISC SDTM datasets:

Reviewer 3: link missing for pharmaversesdtm

``` r
library(admiral)
library(dplyr, warn.conflicts = FALSE)
library(pharmaversesdtm)
library(stringr)
library(magrittr)

# Load sample data
data("dm", package = "pharmaversesdtm")
data("ex", package = "pharmaversesdtm")
data("ds", package = "pharmaversesdtm")
```

ADSL is typically **built from the DM dataset**, removing unnecessary columns and adding treatment variables in one step:

``` r
adsl <- dm %>%
  select(-DOMAIN) %>%
  mutate(
    TRT01P = ARM,
    TRT01A = ACTARM
  )
```

#### **Step 2: Derive Treatment Variables**

Using {admiral}, we extract and standardize treatment dates from the EX dataset:

``` r
ex_ext <- ex %>%
  filter(!is.na(USUBJID)) %>%
  derive_vars_dt(
    dtc = EXSTDTC,
    new_vars_prefix = "EXST"
  ) %>%
  derive_vars_dt(
    dtc = EXENDTC,
    new_vars_prefix = "EXEN"
  )
```

> #### Then merge these dates into ADSL:

Reviewer 3: ```suggestion
Then merge these dates into ADSL:
```

``` r
adsl <- adsl %>%
  derive_var_merged_exist_flag(
    dataset_add = ex,
    by_vars = exprs(STUDYID, USUBJID),
    new_var = SAFFL,
    condition = EXDOSE > 0 | str_detect(EXTRT, "PLACEBO")
  )
```

#### **Step 5: Generate and Save Results**

Finally, we print the results and save them to a CSV file:

``` r
tryCatch({
  print("ADSL creation complete. Showing first few rows:")
  print(head(adsl))
  
  print("\nSummary of ADSL variables:")
  print(summary(adsl))
  
  # Save to a CSV file
  write.csv(adsl, "adsl_output.csv", row.names = FALSE)
  print("\nADSL has been saved to 'adsl_output.csv'")
}, error = function(e) {
  print(paste("An error occurred:", e$message))
})
```

![](images/result1.png)

![](images/result2.png)\

#### **More Details on ADSL Creation**

This is just a **high-level example**; the full process includes deriving death variables, grouping populations, and applying labels. For a deeper dive, check out the [ADSL Implementation Guide](https://cran.r-project.org/web/packages/admiral/vignettes/adsl.html).

![](images/adsl.png)

## **Who Are the Key Players in Pharmaverse, and Do You Need to Use All Packages?**

> [![Source: Getting Started with tidyverse in R](images/tidyverse-packages.png)](https://www.storybench.org/getting-started-with-tidyverse-in-r/)

Reviewer 3: Question: does this image make sense here? the section is about pharmaverse but the image is tidyverse

#### **Key Players in pharmaverse**

-   **Pharma Industry Leaders** – Pharmaceutical companies, CROs, and regulatory bodies driving adoption.

> -   **Regulatory Agencies such as the FDA and EMA, and advisory agencies such as [CDISC](https://www.cdisc.org/), [PHUSE](https://phuse.global/))** – Setting standards for clinical data reporting.

Reviewer 4: This makes it sound like all these are part of the pharmaverse. I would say move this bullet to the bottom and reword as something like "Pharmaverse council and community have outreach to numerous other organisations such as..." and include R Consortium too.

Reviewer 4: I would also call out PHUSE separately as they support Pharmaverse so actively back us. For this bullet you could say `"Pharmaverse is supported by [PHUSE](https://phuse.global/)"`

-   **Open-Source Community** – Developers and passionate individual contributors contributing to pharmaverse packages to improve functionality and compliance.

-   **Organizations and Research Institutes** – Using pharmaverse for clinical trial data management.

> #### **Do You Need to Use All pharmaverse Packages?**

Reviewer 3: ```suggestion
#### **Do You Need to Use All Pharmaverse Packages?**
```

-   No, organizations can select only the packages that fit their needs.

-   Many packages are modular and independent, allowing selective integration.

> -   pharmaverse hosts multiple packages with similar aims, giving users the flexibility to choose what works best for them rather than prescribing a single approach.

Reviewer 3: ```suggestion
-   Pharmaverse hosts multiple packages with similar aims, giving users the flexibility to choose what works best for them rather than prescribing a single approach.
```

> -   Some packages offer SAS-like functionality, easing transitions for SAS users.

Reviewer 4: I would remove this. I rather not make reference to any proprietary software

-   pharmaverse complements [Tidyverse](https://www.tidyverse.org/), allowing organizations to continue using familiar R workflows.

> ## **How pharmaverse Differs from Tidyverse & How to Learn It Effectively**

Reviewer 3: ```suggestion
## **How Pharmaverse Differs from Tidyverse & How to Learn It Effectively**
```

#### **Differences Between pharmaverse and Tidyverse**

> -   pharmaverse is designed specifically for clinical trial data management, while Tidyverse provides general-purpose data science tools.

Reviewer 3: ```suggestion
-   Pharmaverse is designed specifically for clinical trial data management, while Tidyverse provides general-purpose data science tools.
```

> -   pharmaverse is designed specifically for clinical trial data management, while Tidyverse provides general-purpose data science tools.

Reviewer 4: General comment: We usually use "clinical reporting" in pharmaverse site around our scope rather than "clinical trial data management" as it has slightly wider connotations

> -   pharmaverse integrates Tidyverse functions but adds compliance, validation, and reporting features for regulatory submissions.

Reviewer 3: ```suggestion
-   Pharmaverse integrates Tidyverse functions but adds compliance, validation, and reporting features for regulatory submissions.
```

Reviewer 3: also: do we add validation?

> -   Tidyverse focuses on data wrangling and visualization, while pharmaverse extends this to clinical data structuring, validation, and submission-ready formatting.

Reviewer 4: ```suggestion
-   Tidyverse focuses on generic data wrangling and visualization, while pharmaverse extends this to pharma-specific clinical data structuring, reporting, and submission-ready formatting.
```

> #### **How to Learn pharmaverse Effectively**

Reviewer 3: ```suggestion
#### **How to Learn Pharmaverse Effectively**
```

-   [Start with pharmaverse Examples](https://pharmaverse.github.io/examples/) – A curated set of documentation and tutorials.

-   Attend Pharma Industry Webinars and Conferences – Stay updated on new developments through events like [PHUSE events and webinars](https://phuse.global/Events_Calendar), [R/Pharma conferences and events](https://rinpharma.com/), [CDISC events](https://www.cdisc.org/events), [Shiny Gatherings x Pharmaverse webinars](https://www.youtube.com/playlist?list=PLexAKolMzPcpzPAXNU6KlQ_UpIFsuEPCx), etc.

> -   Engage with the Open-Source Community – Contribute to package improvements or discussions. [You can join the pharmaverse community to get started. ](https://pharmaverse.org/support/)

Reviewer 4: ```suggestion
-   Engage with the Open-Source Community – Contribute to package improvements or discussions. [You can join the pharmaverse community to get started.](https://pharmaverse.org/support/)
```

> -   [Explore Example Use Cases](https://pharmaverse.github.io/examples/) – Hands-on practice with clinical trial datasets.

Reviewer 4: This one is mentioned twice

> ## **Getting Started with the pharmaverse**

Reviewer 3: ```suggestion
## **Getting Started with the Pharmaverse**
```

> pharmaverse provides an open-source ecosystem for clinical trial data management, extending Tidyverse with validation, compliance, and regulatory submission capabilities. By following a structured approach from raw data to ADaMs, organizations can enhance efficiency while maintaining data integrity.

Reviewer 3: ```suggestion
Pharmaverse provides an open-source ecosystem for clinical trial data management, extending Tidyverse with validation, compliance, and regulatory submission capabilities. By following a structured approach from raw data to ADaMs, organizations can enhance efficiency while maintaining data integrity.
```

Reviewer 3: But this section kind of repeats what was said above, so could consider cutting?

Reviewer 4: agree

-   Creating ADSL datasets and essential programming steps

-   Key players in Pharmaverse and whether you need all packages

-   How pharmaverse compares to Tidyverse and how to learn it

By the end, you'll have a clear understanding of how Pharmaverse supports clinical trial operations and how to apply these tools in your work.

## **Key Stages of the Clinical Trial Process**

Managing clinical trial data involves multiple stages, each with its own challenges. pharmaverse provides a range of packages for different parts of the process, sometimes even more than one option for the same task, so you can choose what works best for your needs rather than sticking to a single approach

![](images/clinical-trial-process.png)

1.  **Collection** – Initial data handling using:

    -   [**tidyverse**](https://www.tidyverse.org/) for general data manipulation.

    -   [**diffdf**](https://cran.r-project.org/web/packages/diffdf/vignettes/diffdf-basic.html) for tracking differences in datasets.

    -   [**metatools**](https://pharmaverse.github.io/metatools/) for metadata management and transformation.

2.  **Tabulation** – Structuring and validating data with:

>     -   [**timber**](https://rdrr.io/github/atorus-research/timber/) for traceability and logging.

Reviewer: this package is called logrx now

Reviewer: i say this is more a utility package then tabulation package as it can create logs for any package

    -   [**thevalidatoR**](https://github.com/marketplace/actions/r-package-validation-report) for data validation checks.

    -   [**covtracer**](https://github.com/Genentech/covtracer) for ensuring code traceability and reproducibility.

3.  **Analysis Ready** – Preparing data for regulatory-grade analytics:

    -   [**admiral**](https://github.com/pharmaverse/admiral) for standardized data derivations.

    -   [**metacore**](https://cran.r-project.org/web/packages/metacore/index.html) for metadata-driven structures.

>     -   [**Dplyr**](https://dplyr.tidyverse.org/) (part of the tidyverse) for data manipulation

Reviewer: ```suggestion
    -   [**dplyr**](https://dplyr.tidyverse.org/) (part of the tidyverse) for data manipulation
```

Finally, we print the results and save them to a CSV file:

``` r
tryCatch({
  print("ADSL creation complete. Showing first few rows:")
  print(head(adsl))
  
  print("\nSummary of ADSL variables:")
  print(summary(adsl))
  
  # Save to a CSV file
  write.csv(adsl, "adsl_output.csv", row.names = FALSE)
  print("\nADSL has been saved to 'adsl_output.csv'")
}, error = function(e) {
  print(paste("An error occurred:", e$message))
})
```

![](images/result1.png)

![](images/result2.png)\

#### **More Details on ADSL Creation**

This is just a **high-level example**; the full process includes deriving death variables, grouping populations, and applying labels. For a deeper dive, check out the [ADSL Implementation Guide](https://cran.r-project.org/web/packages/admiral/vignettes/adsl.html).

![](images/adsl.png)

## **Who Are the Key Players in Pharmaverse, and Do You Need to Use All Packages?**

[![Source: Getting Started with tidyverse in R](images/tidyverse-packages.png)](https://www.storybench.org/getting-started-with-tidyverse-in-r/)

#### **Key Players in pharmaverse**

-   **Pharma Industry Leaders** – Pharmaceutical companies, CROs, and regulatory bodies driving adoption.

> -   **Regulatory Agencies (FDA, [CDISC](https://www.cdisc.org/), [PHUSE](https://phuse.global/))** – Setting standards for clinical data reporting.

Reviewer: ummm.....CDISC and PHUSE are not regulatory agencies...maybe advisory agencies is better word for them?

### **Resources**

-   This blog post was based on this presentation by Sunil Gupta: [R and pharmaverse: The New Frontier for Today’s Statistical Programmers](https://phuse.s3.eu-central-1.amazonaws.com/Archive/2024/SDE/US/Mississauga/PRE_Mississauga05.pdf)

-   [R-Guru Resource Hub for Rapid R Learning](https://www.lexjansen.com/phuse-us/2024/pd/PAP_PD05.pdf)

-   [Explore more posts in the pharmaverse blog](https://rinpharma.shinyapps.io/tidyCDISC/)

-   [Subscribe to the pharmaverse newsletter](https://forms.office.com/pages/responsepage.aspx?id=xeEJLj1cykuXxFc6VpX1UAPF0grflaRJu8z6VC7-hy5UMTE0M0lXR1JON1Q0MkRIRlA1TlZSNDhUSC4u&route=shorturl)

<!--------------- appendices go here ----------------->

> ```{r, echo=FALSE}

Reviewer: ```suggestion
```{r, echo=FALSE, message = FALSE}
```

-   The **pharmaverse** provides multiple table-making packages, such as [chevron](https://insightsengineering.github.io/chevron/latest-tag/), [Tplyr](https://github.com/atorus-research/Tplyr), [pharmaRTF](https://atorus-research.github.io/pharmaRTF/), [gtsummary](https://www.danieldsjoberg.com/gtsummary/), [cards](https://insightsengineering.github.io/cards/latest-tag/), [tfrmt](https://gsk-biostatistics.github.io/tfrmt/), and [tidytlg](https://pharmaverse.github.io/tidytlg/main/). More tools are listed on the [TLGs page](https://pharmaverse.org/e2eclinical/tlg/).
-   [**xportr**](https://github.com/atorus-research/xportr) – CDISC-compliant dataset export.
-   [**pkglite**](https://merck.github.io/pkglite/) – Package management and tracking.
-   [**metacore**](https://github.com/atorus-research/metacore), [**metatools**](https://pharmaverse.github.io/metatools/) – For standardized metadata structures and validation.
-   [**sdtmchecks**](https://github.com/pharmaverse/sdtmchecks), [**datacutr**](https://cran.r-project.org/web/packages/datacutr/index.html) – Ensuring data integrity and regulatory compliance.

> 

Reviewer: Can we add https://github.com/pharmaverse/logrx to this list for logging R scripts

Reviewer 3: done

By using these tools, organizations can optimize their data pipeline, ensuring clinical data is well-structured and ready for regulatory submission with ease.

## **Example: Creating ADSL**

Building an ADSL dataset involves several key steps, from reading in data to deriving treatment variables and population flags. While these steps apply regardless of the tools used, **pharmaverse packages like {admiral} simplify the process** with functions designed for CDISC-compliant datasets.

This example is based on the [ADSL template](https://cran.r-project.org/web/packages/admiral/vignettes/adsl.html), which provides a structured approach to creating an ADSL dataset.

#### **Step 1: Read in Data**

To begin, clinical trial datasets such as **DM, EX, DS, AE, and LB** are loaded. The [{pharmaversesdtm}](https://github.com/pharmaverse/pharmaversesdtm) package provides sample CDISC SDTM datasets:

``` r
library(admiral)
library(dplyr, warn.conflicts = FALSE)
library(pharmaversesdtm)
library(stringr)

> library(magrittr)

Reviewer: this can be removed

Reviewer 3: done

``` r
ds_ext <- ds %>%
  filter(!is.na(DSSTDTC)) %>%
  derive_vars_dt(
    dtc = DSSTDTC,
    new_vars_prefix = "DSST"
  )

adsl <- adsl %>%
  derive_vars_merged(
    dataset_add = ds_ext,
    by_vars = exprs(STUDYID, USUBJID),
    new_vars = exprs(EOSDT = DSSTDT),
    filter_add = DSCAT == "DISPOSITION EVENT" & DSDECOD != "SCREEN FAILURE"
  )
```

#### **Step 4: Assign Population Flags**

For safety population (`SAFFL`), we check if the patient received a treatment dose:

``` r
adsl <- adsl %>%
  derive_var_merged_exist_flag(
    dataset_add = ex,
    by_vars = exprs(STUDYID, USUBJID),
    new_var = SAFFL,
    condition = EXDOSE > 0 | str_detect(EXTRT, "PLACEBO")
  )
```

#### **Step 5: Generate and Save Results**

Finally, we print the results and save them to a CSV file:

``` r
tryCatch({
  print("ADSL creation complete. Showing first few rows:")
  print(head(adsl))
  
  print("\nSummary of ADSL variables:")
  print(summary(adsl))
  
  # Save to a CSV file
  write.csv(adsl, "adsl_output.csv", row.names = FALSE)
  print("\nADSL has been saved to 'adsl_output.csv'")
}, error = function(e) {
  print(paste("An error occurred:", e$message))
})
```
**Results**

``` r

> [1] "ADSL creation complete. Showing first few rows:"

Reviewer: can we just use reactable to make nice scrollabe dataframe for this?

                                                                                                NA's   :52   
    TRT01P             TRT01A              EOSDT               SAFFL          
 Length:306         Length:306         Min.   :2012-09-01   Length:306        
 Class :character   Class :character   1st Qu.:2013-05-16   Class :character  
 Mode  :character   Mode  :character   Median :2013-09-30   Mode  :character  
                                       Mean   :2013-10-15                     
                                       3rd Qu.:2014-03-20                     
                                       Max.   :2015-03-05                     
                                       NA's   :52                             
[1] "\nADSL has been saved to 'adsl_output.csv'"
```

#### **More Details on ADSL Creation**

This is just a **high-level example**; the full process includes deriving death variables, grouping populations, and applying labels. For a deeper dive, check out the [ADSL Implementation Guide](https://cran.r-project.org/web/packages/admiral/vignettes/adsl.html).

> ![](images/adsl.png)

Reviewer: Umm...I think this image can just be dropped - the link is enough

Reviewer 3: dropped

-   **The pharmaverse community collaborates with organizations like the FDA, EMA, R Consortium, and CDISC** to align with industry standards and best practices for clinical data reporting.

### **Do You Need to Use All Pharmaverse Packages?**

-   No, organizations can select only the packages that fit their needs.

-   Many packages are modular and independent, allowing selective integration.

-   Pharmaverse hosts multiple packages with similar aims, giving users the flexibility to choose what works best for them rather than prescribing a single approach.

-   Pharmaverse complements [Tidyverse](https://www.tidyverse.org/), allowing organizations to continue using familiar R workflows.

## **How Pharmaverse Differs from Tidyverse & How to Learn It Effectively**

#### **Differences Between pharmaverse and Tidyverse**

-   Pharmaverse is designed specifically for clinical reporting, while Tidyverse provides general-purpose data science tools.

-   Pharmaverse integrates Tidyverse functions but adds compliance, validation, and reporting features for regulatory submissions.

> -   Tidyverse focuses on generic data wrangling and visualization, while pharmaverse extends this to pharma-specific clinical data structuring, reporting, and submission-ready formatting.

Reviewer 3: done

Pharmaverse provides an open-source ecosystem for clinical reporting, extending Tidyverse with validation, compliance, and regulatory submission capabilities. By following a structured approach from raw data to ADaMs, organizations can enhance efficiency while maintaining data integrity.

- You can [start with Pharmaverse Examples](https://pharmaverse.github.io/examples/) – A curated set of documentation and tutorials.

- Attend Pharma Industry Webinars and Conferences – Stay updated on new developments through events like [PHUSE events and webinars](https://phuse.global/Events_Calendar), [R/Pharma conferences and events](https://rinpharma.com/), [CDISC events](https://www.cdisc.org/events), [Shiny Gatherings x Pharmaverse webinars](https://www.youtube.com/playlist?list=PLexAKolMzPcpzPAXNU6KlQ_UpIFsuEPCx), etc.

-  Engage with the Open-Source Community – Contribute to package improvements or discussions. [You can join the pharmaverse community to get started.](join the pharmaverse](https://join.slack.com/t/pharmaverse/shared_invite/zt-yv5atkr4-Np2ytJ6W_QKz_4Olo7Jo9A)).

- Explore packages on the [pharmaverse website](https://pharmaverse.org).

> - Try implementing an ADSL dataset using [CRAN Admiral ADSL](https://cran.r-project.org/web/packages/admiral/vignettes/adsl.html). 

Reviewer: What is CRAN Admiral ADSL?

Reviewer 3: renamed
